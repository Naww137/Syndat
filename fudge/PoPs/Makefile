# <<BEGIN-copyright>>
# Copyright 2022, Lawrence Livermore National Security, LLC.
# See the top-level COPYRIGHT file for details.
# 
# SPDX-License-Identifier: BSD-3-Clause
# <<END-copyright>>

SUB_DIRS = quantities chemicalElements families decays atomic Test

.PHONY: clean realclean doTarget

default:

clean:
	rm -f *.pyc
	$(MAKE) doTarget MODE=clean

realclean: clean
	$(MAKE) doTarget MODE=realclean

doTarget:
	SAVED_PWD=`pwd`; \
	for directory in $(SUB_DIRS); do cd $$directory; $(MAKE) $(MODE); cd $$SAVED_PWD; done
