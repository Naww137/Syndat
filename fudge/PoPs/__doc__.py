# <<BEGIN-copyright>>
# Copyright 2022, Lawrence Livermore National Security, LLC.
# See the top-level COPYRIGHT file for details.
# 
# SPDX-License-Identifier: BSD-3-Clause
# <<END-copyright>>

"""
This is the 'Properties of Particles' or PoPs package. It defines a set of Python classes
for storing intrinsic particle properties like charge, mass, spin, etc.
"""
