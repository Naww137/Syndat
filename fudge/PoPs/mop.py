# <<BEGIN-copyright>>
# Copyright 2022, Lawrence Livermore National Security, LLC.
# See the top-level COPYRIGHT file for details.
# 
# SPDX-License-Identifier: BSD-3-Clause
# <<END-copyright>>

__doc__ = """
This module contains classes to handle a MoP (Map of PoPs) instance.
"""

import os
import re

from LUPY import ancestry as ancestryModule
from . import database as databaseModule

class FormatVersion :

    version_0_1 = '0.1'

    allowed = ( version_0_1, )

class Mop(ancestryModule.AncestryIO):

    moniker = 'mop'

    def __init__(self, name, path = './', format=FormatVersion.version_0_1):

        """
        Constructor for a Mop instance.

        :param name:            The name of the library.
        :param path:            The path to the mop file if read in from disk.
        :param format:          The format version for the mop.
        """

        ancestryModule.AncestryIO.__init__(self)

        if( not( isinstance( name, str ) ) ) : raise TypeError( 'Library must be a string.' )
        self.__name = name 

        if( not( isinstance( path, str ) ) ) : raise TypeError( 'Library must be a string.' )
        self.__path = path

        if( format not in FormatVersion.allowed ) : raise ValueError( 'Invalid format "%s."' % format )
        self.__format = format

        self.__entries = []
        self.__pops = databaseModule.Database( 'Internal', '0.1' )

    def __len__( self ) :
        """Returns the number of entries in self."""

        return( len( self__entries ) )

    def __getitem__( self, index ) :
        """
        Returns the (index-1)^th item of self.

        :param index:       The index of the iter to return.
        """

        return( self.__entries[index] )

    @property
    def format( self ) :
        """Returns the format self is stored as."""

        return( self.__format )

    @property
    def name( self ) :
        """Returns self's name."""

        return( self.__name )

    @property
    def path( self ) :
        """Returns the path to self. If the data in self was read from disk, this is the path to the file."""

        return( self.__path )

    @property
    def fileName( self ) :
        """Same the path method."""

        return( self.__path )

    def append( self, entry ) :
        """Appends *entry* to the end of self's list of entries."""

        if( not( isinstance( entry, EntryBase ) ) ) : raise TypeError( 'Invalid instance to add.' )

        self.__entries.append( entry )
        entry.setAncestor( self )

    def fetch( self, re_pid, readPoPs ) :
        """
        Returns a list of items. Each item contains information about a PoPs database where one or more matches to *re_pid* exists. 
        Each item contains the list [ ids, all_ids, fileName, pops ]) where:

            -) ids:         The list of PoPs ids in a PoPs database that matches *re_pid*,
            -) all_ids:     The list of all PoPs ids in the PoPs database,
            -) fileName:    The file path to the PoPs database,
            -) pops:        An instance of the PoPs database if *readPoPs* is *True* and *None* otherwise.

        The *ids* are a list of str ids (e.g., [ 'Pu234', 'Pu234e_1' ]).  *re_pid* can be a str instance or a *re.compile* instance.
        Note, even if *readPoPs* is False, the returned pops will not be None if it had previously been read in.
        
        :param re_pid:          A str or a *re.compile* pattern representing the ids to match.
        :param readPoPs:        If *True*, the pops file is read and returned as the fourth element of each returned item.
        """

        matches = []
        for entry in self.__entries : matches += entry.fetch( re_pid, readPoPs )

        return( matches )

    def toXML_strList( self, indent = "", **kwargs ) :
        """
        Returns a list of str instances representing the XML lines of self.

        :param indent:          The amount of indentation for each line. Child nodes and text may be indented more.
        :param **kwargs:        A keyword list.

        :return:                List of str instances representing the XML lines of self.
        """

        indent2 = indent + kwargs.get( 'incrementalIndent', '  ' )

        format = kwargs.get( 'format', FormatVersion.version_0_1 )
        if( format not in FormatVersion.allowed ) : raise TypeError( 'Invalid format = "%s".' % format )

        XML_list = [ '%s<%s name="%s" format="%s">' % ( indent, self.moniker, self.name, format ) ]
        for entry in self.__entries : XML_list += entry.toXML_strList( indent2, **kwargs )
        XML_list[-1] +=  '</%s>' % self.moniker

        return( XML_list )

    @staticmethod
    def parseNode( node, path ) :
        """
        Creates a Mop instance from an XML mop element.

        :param node:        XML element to parse.

        :return:            Mop instance.
        """

        if( node.tag != Mop.moniker ) : raise TypeError( 'Invalid node name. Expected "%s" got "%s".' % ( Mop.moniker, node.tag ) )

        format = node.get( 'format', None )
        if( format is None ) : raise ValueError( 'Mop node does not have a "format" attribute.' )

        name = node.get( 'name', None )
        if( name is None ) : raise ValueError( 'Mop node does not have "name" attribute.' )

        mop = Mop( name, path = path )

        settings = { 'format' : format }
        for child in node :
            if( child.tag == Import.moniker ) :
                mop.append( Import.parseNode( settings, child ) )
            elif( child.tag == PoPs.moniker ) :
                mop.append( PoPs.parseNode( settings, child ) )
            else :
                raise ValueError( 'Invalid child tag "%s" for mop file.' % child.tag )

        return( mop )

class EntryBase(ancestryModule.AncestryIO) :
    """
    Base class for mop entries.
    """

    def __init__( self, path ) :
        """
        Base constructor for mop entries.

        :param path:            Path to the file for the entry.
        """

        ancestryModule.AncestryIO.__init__(self)

        if( not( isinstance( path, str ) ) ) : raise TypeError( 'Path must be a string.' )
        self.__path = path

    @property
    def fileName( self ) :
        """Returns the absolute path to the file pointed to by the 'path' attribute."""

        filename = self.path
        if( filename != os.sep ) : filename = os.path.join( os.path.dirname( os.path.realpath( self.ancestor.fileName ) ), filename )
        return( filename )

    @property
    def path( self ) :
        """Returns the path of the mop file. This may be absolute or relative."""

        return( self.__path )

class PoPs(EntryBase):
    """
    Class for storing a GNDS mop's PoPs instance.
    """

    moniker = 'pops'

    def __init__( self, path, ids ) :
        """
        Constructor for a mop's PoPs instance.

        :param path:            Path to the GNDS PoPs.database file.
        :param ids:             The list of particles ids in the PoPs.database file.
        """

        EntryBase.__init__(self, path)
        self.__pops = None
        self.__ids = ids

    @property
    def ids( self ) :
        """Returns the list of ids."""

        return( self.__ids )

    @property
    def pops( self ) :
        """Returns a reference to the PoPs.database instance. If the PoPs file has not be read in (see method *readIfNeeded*), None is returned."""

        return( self.__pops )

    def fetch( self, re_pid, readPoPs ) :
        """
        Returns the list [ ids, all_ids, fileName, pops ] if matches to *re_pid* are found in *self* and an empty list otherwise. 

            -) ids:         The list of PoPs ids in the PoPs database that matches *re_pid*,
            -) all_ids:     The list of all PoPs ids in the PoPs database,
            -) fileName:    The path to the PoPs database,
            -) pops:        An instance of the PoPs database if *readPoPs* is *True* and *None* otherwise.

        The *ids* are a list of str ids (e.g., [ 'Pu234', 'Pu234e_1' ]). *re_pid* can be a str instance or a *re.compile* instance.
        
        :param re_pid:          A str or *re.compile* pattern representing ids to match.
        :param readPoPs:        If *True*, the pops file is read and returned as the fourth element.
        """

        if( isinstance( re_pid, str ) ) : re_pid = re.compile( re_pid )
        matches = [ pid for pid in self.ids if re_pid.fullmatch( pid ) ]

        if( len( matches ) > 0 ) :
            matches = [ [ matches, self.ids, self.fileName, self.readIfNeeded( readPoPs ) ] ]

        return( matches )

    def readIfNeeded( self, readPoPs = True ) :
        """
        Reads in the PoPs.database referenced by the *path* member if it has not been read in by a prior call and *readPoPs* 
        is True (default). Also, returns the value of *self.__pops*. The returned value will only be *None* if *readPoPs* 
        is *False* and all prior calls to this method had "*readPoPs* = False".

        :param readPoPs:        If *True*, the pops file is read.
        :return:                The value if *self.__pops*.
        """

        if( ( self.__pops is None ) and readPoPs ) :
            self.__pops = databaseModule.read( self.fileName )
            self.__pops.setAncestor( self )

        return( self.__pops )

    def toXML_strList( self, indent = '', **kwargs ) :
        """
        Returns a list of str instances representing the XML lines of self.

        :param indent:          The amount of indentation for each line. Child nodes and text may be indented more.
        :param **kwargs:        A keyword list.

        :return:                List of str instances representing the XML lines of self.
        """

        return( [ '%s<%s path="%s">' % ( indent, self.moniker, self.path ) + ' '.join( self.__ids ) + '</%s>' % self.moniker ] )

    @classmethod
    def parseNodeUsingClass(settings, node):
        """
        Creates an Import instance from an XML import element.

        :param settings:        Information from parent map instances needed to parse the XML element.
        :param node:            XML element to parse.

        :return:                Import instance.
        """

        if( node.tag != PoPs.moniker ) : raise TypeError( 'Invalid node name. Expected "%s" got "%s".' % ( PoPs.moniker, node.tag ) )

        path = node.get( 'path', None )
        ids = node.text.split( )

        pops = PoPs( path, ids )

        return( pops )

class Import(EntryBase):
    """
    Class for storing a GNDS mop's Import instance.
    """

    moniker = 'import'

    def __init__( self, path ) :
        """
        Constructor for a mop's import instance.

        :param path:            Path to the GNDS mop file.
        """

        EntryBase.__init__(self, path)
        self.__mop = None

    @property
    def mop( self ) :
        """Returns a reference to self.__mop."""

        return( self.__mop )

    def fetch( self, re_pid, readPoPs ) :
        """
        See documentation for Mop.fetch.
        
        :param re_pid:          A str or *re* pattern representing ids to match.
        :param readPoPs:        If *True*, the pops file is read and returned as the fourth element of each returned item.
        """

        self.readIfNeeded( )
        return( self.__mop.fetch( re_pid, readPoPs ) )

    def readIfNeeded( self ) :
        """
        Reads in the PoPs.database referenced by the *path* member if it has not been read in.
        """

        if( self.__mop is None ) :
            self.__mop = Mop.readXML_file( self.fileName )
            self.__mop.setAncestor( self )

    def toXML_strList( self, indent = '', **kwargs ) :
        """
        Returns a list of str instances representing the XML lines of self.

        :param indent:          The amount of indentation for each line. Child nodes and text may be indented more.
        :param **kwargs:        A keyword list.

        :return:                List of str instances representing the XML lines of self.
        """

        return( [ '%s<%s path="%s"/>' % ( indent, self.moniker, self.path ) ] )

    @staticmethod
    def parseNode( settings, node ) :
        """
        Creates an Import instance from an XML import element.

        :param settings:        Information from parent map instances needed to parse the XML element.
        :param node:            XML element to parse.

        :return:                Import instance.
        """

        if( node.tag != Import.moniker ) : raise TypeError( 'Invalid node name. Expected "%s" got "%s".' % ( Import.moniker, node.tag ) )

        path = node.get( 'path', None )

        _import = Import( path )

        return( _import )
