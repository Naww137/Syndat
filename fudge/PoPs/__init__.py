# <<BEGIN-copyright>>
# Copyright 2022, Lawrence Livermore National Security, LLC.
# See the top-level COPYRIGHT file for details.
# 
# SPDX-License-Identifier: BSD-3-Clause
# <<END-copyright>>

"""
Top-level __init__ for the PoPs package.
"""

POPS_MAJORVERSION = 0
POPS_MINORVERSION = 1
POPS_PATCHLEVEL = 0
__version__ = '%s.%s.%s' % ( POPS_MAJORVERSION, POPS_MINORVERSION, POPS_PATCHLEVEL )

from .__doc__ import __doc__
