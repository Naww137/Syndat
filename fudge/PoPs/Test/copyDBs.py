#! /usr/bin/env python3

# <<BEGIN-copyright>>
# Copyright 2022, Lawrence Livermore National Security, LLC.
# See the top-level COPYRIGHT file for details.
# 
# SPDX-License-Identifier: BSD-3-Clause
# <<END-copyright>>

from PoPs import database as databaseModule

fOut = open( 'Outputs/copyDBs.py.copy.out', 'w' )

def _copy( fileName ) :

    fIn = open( fileName )
    fOut.writelines( fIn.readlines( ) )
    fIn.close( )

    database1 = databaseModule.read( fileName )
    database2 = database1.copy( )
    print( database2.toXML( ) )

_copy( 'Answers/database.py.out' )
_copy( 'Answers/database3.py.out' )
_copy( 'Answers/database4.py.out' )

fOut.close( )
