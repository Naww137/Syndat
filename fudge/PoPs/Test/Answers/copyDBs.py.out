<PoPs name="test" version="1.2.3" format="0.1">
  <aliases>
    <metaStable id="Am242_m1" pid="Am242_e2" metaStableIndex="1"/>
    <alias id="electron" pid="e-"/>
    <alias id="e+" pid="e-_anti"/>
    <alias id="positron" pid="e-_anti"/></aliases>
  <gaugeBosons>
    <gaugeBoson id="photon">
      <mass>
        <double label="base" value="0." unit="amu"/></mass>
      <spin>
        <fraction label="base" value="1" unit="hbar"/></spin>
      <parity>
        <integer label="base" value="1"/></parity>
      <charge>
        <integer label="base" value="0" unit="e"/></charge>
      <halflife>
        <string label="base" value="stable" unit="s"/></halflife></gaugeBoson></gaugeBosons>
  <leptons>
    <lepton id="e-" generation="electronic">
      <mass>
        <double label="base" value="5.4857990907e-4" unit="amu"/></mass>
      <spin>
        <fraction label="base" value="1/2" unit="hbar"/></spin>
      <parity>
        <integer label="base" value="1"/></parity>
      <charge>
        <integer label="base" value="-1" unit="e"/></charge>
      <halflife>
        <string label="base" value="stable" unit="s"/></halflife></lepton>
    <lepton id="e-_anti" generation="electronic">
      <mass>
        <double label="base" value="5.4857990907e-4" unit="amu"/></mass>
      <spin>
        <fraction label="base" value="1/2" unit="hbar"/></spin>
      <parity>
        <integer label="base" value="-1"/></parity>
      <charge>
        <integer label="base" value="1" unit="e"/></charge>
      <halflife>
        <string label="base" value="stable" unit="s"/></halflife></lepton>
    <lepton id="mu" generation="muonic">
      <mass>
        <double label="base" value="0.1134289267" unit="amu"/></mass>
      <spin>
        <fraction label="base" value="1/2" unit="hbar"/></spin>
      <parity>
        <integer label="base" value="1"/></parity>
      <charge>
        <integer label="base" value="-1" unit="e"/></charge>
      <halflife>
        <double label="base" value="2.1969811e-6" unit="s"/></halflife></lepton></leptons>
  <baryons>
    <baryon id="n">
      <mass>
        <double label="base" value="1.00866491588" unit="amu"/></mass>
      <spin>
        <fraction label="base" value="1/2" unit="hbar"/></spin>
      <parity>
        <integer label="base" value="1"/></parity>
      <charge>
        <integer label="base" value="0" unit="e"/></charge>
      <halflife>
        <double label="base" value="881.5" unit="s"/></halflife></baryon>
    <baryon id="n_anti">
      <mass>
        <double label="base" value="1.00866491588" unit="amu"/></mass>
      <spin>
        <fraction label="base" value="1/2" unit="hbar"/></spin>
      <parity>
        <integer label="base" value="1"/></parity>
      <charge>
        <integer label="base" value="0" unit="e"/></charge>
      <halflife>
        <double label="base" value="881.5" unit="s"/></halflife></baryon>
    <baryon id="p">
      <mass>
        <double label="base" value="1.00727646681" unit="amu"/></mass>
      <spin>
        <fraction label="base" value="1/2" unit="hbar"/></spin>
      <parity>
        <integer label="base" value="1"/></parity>
      <charge>
        <integer label="base" value="1" unit="e"/></charge>
      <halflife>
        <string label="base" value="stable" unit="s"/></halflife></baryon>
    <baryon id="p_anti">
      <mass>
        <double label="base" value="1.00727646681" unit="amu"/></mass>
      <spin>
        <fraction label="base" value="1/2" unit="hbar"/></spin>
      <parity>
        <integer label="base" value="1"/></parity>
      <charge>
        <integer label="base" value="1" unit="e"/></charge>
      <halflife>
        <string label="base" value="stable" unit="s"/></halflife></baryon></baryons>
  <chemicalElements>
    <chemicalElement symbol="O" Z="8" name="Oxygen">
      <isotopes>
        <isotope symbol="O16" A="16">
          <nuclides>
            <nuclide id="O16">
              <mass>
                <double label="base" value="15.9949146196" unit="amu"/></mass>
              <nucleus id="o16" index="0">
                <energy>
                  <double label="base" value="0." unit="keV"/></energy></nucleus></nuclide>
            <nuclide id="O16_e1">
              <nucleus id="o16_e1" index="1">
                <energy>
                  <double label="base" value="6049.4" unit="keV"/></energy></nucleus></nuclide>
            <nuclide id="O16_e2">
              <nucleus id="o16_e2" index="2">
                <energy>
                  <double label="base" value="6129.893" unit="keV"/></energy></nucleus></nuclide>
            <nuclide id="O16_e3">
              <nucleus id="o16_e3" index="3">
                <energy>
                  <double label="base" value="6917.1" unit="keV"/></energy></nucleus></nuclide></nuclides></isotope>
        <isotope symbol="O17" A="17">
          <nuclides>
            <nuclide id="O17">
              <mass>
                <double label="base" value="15.9949146196" unit="amu"/></mass>
              <nucleus id="o17" index="0">
                <energy>
                  <double label="base" value="0." unit="keV"/></energy></nucleus></nuclide>
            <nuclide id="O17_e1">
              <nucleus id="o17_e1" index="1">
                <energy>
                  <double label="base" value="6049.4" unit="keV"/></energy></nucleus></nuclide>
            <nuclide id="O17_e2">
              <nucleus id="o17_e2" index="2">
                <energy>
                  <double label="base" value="6129.893" unit="keV"/></energy></nucleus></nuclide>
            <nuclide id="O17_e3">
              <nucleus id="o17_e3" index="3">
                <energy>
                  <double label="base" value="6917.1" unit="keV"/></energy></nucleus></nuclide></nuclides></isotope></isotopes></chemicalElement>
    <chemicalElement symbol="Am" Z="95" name="Americium">
      <isotopes>
        <isotope symbol="Am242" A="242">
          <nuclides>
            <nuclide id="Am242">
              <mass>
                <double label="base" value="242.059549159" unit="amu"/></mass>
              <spin>
                <fraction label="base" value="1" unit="hbar"/></spin>
              <parity>
                <integer label="base" value="-1"/></parity>
              <halflife>
                <double label="base" value="16.02" unit="h"/></halflife>
              <nucleus id="am242" index="0">
                <energy>
                  <double label="base" value="0." unit="keV"/></energy></nucleus></nuclide>
            <nuclide id="Am242_e1">
              <spin>
                <fraction label="base" value="0" unit="hbar"/></spin>
              <parity>
                <integer label="base" value="-1"/></parity>
              <nucleus id="am242_e1" index="1">
                <energy>
                  <double label="base" value="44.092" unit="keV"/></energy></nucleus></nuclide>
            <nuclide id="Am242_e2">
              <spin>
                <fraction label="base" value="5" unit="hbar"/></spin>
              <parity>
                <integer label="base" value="-1"/></parity>
              <halflife>
                <double label="base" value="141." unit="yr"/></halflife>
              <nucleus id="am242_e2" index="2">
                <energy>
                  <double label="base" value="48.6" unit="keV"/></energy></nucleus></nuclide>
            <nuclide id="Am242_e3">
              <spin>
                <fraction label="base" value="3" unit="hbar"/></spin>
              <parity>
                <integer label="base" value="-1"/></parity>
              <nucleus id="am242_e3" index="3">
                <energy>
                  <double label="base" value="52.7" unit="keV"/></energy></nucleus></nuclide></nuclides></isotope></isotopes></chemicalElement></chemicalElements></PoPs>
<PoPs name="test" version="1.2.3" format="0.1">
  <gaugeBosons>
    <gaugeBoson id="photon"></gaugeBoson></gaugeBosons>
  <leptons>
    <lepton id="e-" generation="electronic"></lepton>
    <lepton id="mu-_anti" generation="muonic"></lepton></leptons>
  <baryons>
    <baryon id="n_anti"></baryon>
    <baryon id="p"></baryon></baryons>
  <chemicalElements>
    <chemicalElement symbol="N" Z="7" name="Nitrogen">
      <isotopes>
        <isotope symbol="N15" A="15">
          <nuclides>
            <nuclide id="N15_e5">
              <nucleus id="n15_e5" index="5"></nucleus></nuclide>
            <nuclide id="N15_e7">
              <nucleus id="n15_e7" index="7"></nucleus></nuclide></nuclides></isotope></isotopes></chemicalElement>
    <chemicalElement symbol="O" Z="8" name="Oxygen">
      <isotopes>
        <isotope symbol="O16" A="16">
          <nuclides>
            <nuclide id="O16_e12">
              <nucleus id="o16_e12" index="12"></nucleus></nuclide></nuclides></isotope></isotopes></chemicalElement>
    <chemicalElement symbol="Pu" Z="94" name="Plutonium">
      <isotopes>
        <isotope symbol="Pu237" A="237"></isotope>
        <isotope symbol="Pu238" A="238">
          <nuclides>
            <nuclide id="Pu238">
              <nucleus id="pu238" index="0"></nucleus></nuclide>
            <nuclide id="Pu238_e2">
              <nucleus id="pu238_e2" index="2"></nucleus></nuclide></nuclides></isotope>
        <isotope symbol="Pu239" A="239">
          <nuclides>
            <nuclide id="Pu239_e4">
              <nucleus id="pu239_e4" index="4"></nucleus></nuclide></nuclides></isotope></isotopes></chemicalElement></chemicalElements></PoPs>
<PoPs name="test" version="1.2.3" format="0.1">
  <aliases>
    <metaStable id="Am242_m1" pid="Am242_e2" metaStableIndex="1"/>
    <alias id="electron" pid="e-"/>
    <alias id="e+" pid="e-_anti"/>
    <alias id="positron" pid="e-_anti"/>
    <alias id="d" pid="h2"/>
    <alias id="t" pid="h3"/></aliases>
  <gaugeBosons>
    <gaugeBoson id="photon">
      <mass>
        <double label="base" value="0." unit="amu"/></mass>
      <spin>
        <fraction label="base" value="1" unit="hbar"/></spin>
      <parity>
        <integer label="base" value="1"/></parity>
      <charge>
        <integer label="base" value="0" unit="e"/></charge>
      <halflife>
        <string label="base" value="stable" unit="s"/></halflife></gaugeBoson></gaugeBosons>
  <leptons>
    <lepton id="e-" generation="electronic">
      <mass>
        <double label="base" value="5.4857990907e-4" unit="amu"/></mass>
      <spin>
        <fraction label="base" value="1/2" unit="hbar"/></spin>
      <parity>
        <integer label="base" value="1"/></parity>
      <charge>
        <integer label="base" value="-1" unit="e"/></charge>
      <halflife>
        <string label="base" value="stable" unit="s"/></halflife></lepton>
    <lepton id="e-_anti" generation="electronic">
      <mass>
        <double label="base" value="5.4857990907e-4" unit="amu"/></mass>
      <spin>
        <fraction label="base" value="1/2" unit="hbar"/></spin>
      <parity>
        <integer label="base" value="-1"/></parity>
      <charge>
        <integer label="base" value="1" unit="e"/></charge>
      <halflife>
        <string label="base" value="stable" unit="s"/></halflife></lepton>
    <lepton id="mu" generation="muonic">
      <mass>
        <double label="base" value="0.1134289267" unit="amu"/></mass>
      <spin>
        <fraction label="base" value="1/2" unit="hbar"/></spin>
      <parity>
        <integer label="base" value="1"/></parity>
      <charge>
        <integer label="base" value="-1" unit="e"/></charge>
      <halflife>
        <double label="base" value="2.1969811e-6" unit="s"/></halflife></lepton></leptons>
  <baryons>
    <baryon id="n">
      <mass>
        <double label="base" value="1.00866491588" unit="amu"/></mass>
      <spin>
        <fraction label="base" value="1/2" unit="hbar"/></spin>
      <parity>
        <integer label="base" value="1"/></parity>
      <charge>
        <integer label="base" value="0" unit="e"/></charge>
      <halflife>
        <double label="base" value="881.5" unit="s"/></halflife></baryon>
    <baryon id="n_anti">
      <mass>
        <double label="base" value="1.00866491588" unit="amu"/></mass>
      <spin>
        <fraction label="base" value="1/2" unit="hbar"/></spin>
      <parity>
        <integer label="base" value="1"/></parity>
      <charge>
        <integer label="base" value="0" unit="e"/></charge>
      <halflife>
        <double label="base" value="881.5" unit="s"/></halflife></baryon>
    <baryon id="p">
      <mass>
        <double label="base" value="1.00727646681" unit="amu"/></mass>
      <spin>
        <fraction label="base" value="1/2" unit="hbar"/></spin>
      <parity>
        <integer label="base" value="1"/></parity>
      <charge>
        <integer label="base" value="1" unit="e"/></charge>
      <halflife>
        <string label="base" value="stable" unit="s"/></halflife></baryon>
    <baryon id="p_anti">
      <mass>
        <double label="base" value="1.00727646681" unit="amu"/></mass>
      <spin>
        <fraction label="base" value="1/2" unit="hbar"/></spin>
      <parity>
        <integer label="base" value="1"/></parity>
      <charge>
        <integer label="base" value="1" unit="e"/></charge>
      <halflife>
        <string label="base" value="stable" unit="s"/></halflife></baryon></baryons>
  <chemicalElements>
    <chemicalElement symbol="H" Z="1" name="Hydrogen">
      <isotopes>
        <isotope symbol="H1" A="1">
          <nuclides>
            <nuclide id="H1">
              <mass>
                <double label="base" value="1.00782503207" unit="amu"/></mass>
              <nucleus id="h1" index="0">
                <mass>
                  <double label="base" value="1.00727646688" unit="amu"/></mass>
                <spin>
                  <fraction label="base" value="1/2" unit="hbar"/></spin>
                <parity>
                  <integer label="base" value="1"/></parity>
                <charge>
                  <integer label="base" value="1" unit="e"/></charge>
                <halflife>
                  <string label="base" value="stable" unit="s"/></halflife>
                <energy>
                  <double label="base" value="0." unit="keV"/></energy></nucleus></nuclide></nuclides></isotope>
        <isotope symbol="H2" A="2">
          <nuclides>
            <nuclide id="H2">
              <mass>
                <double label="base" value="2.01410177785" unit="amu"/></mass>
              <nucleus id="h2" index="0">
                <mass>
                  <double label="base" value="2.01355319775" unit="amu"/></mass>
                <parity>
                  <integer label="base" value="1"/></parity>
                <charge>
                  <integer label="base" value="1" unit="e"/></charge>
                <halflife>
                  <string label="base" value="stable" unit="s"/></halflife>
                <energy>
                  <double label="base" value="0." unit="keV"/></energy></nucleus></nuclide></nuclides></isotope>
        <isotope symbol="H3" A="3">
          <nuclides>
            <nuclide id="H3">
              <mass>
                <double label="base" value="3.01604927767" unit="amu"/></mass>
              <nucleus id="h3" index="0">
                <mass>
                  <double label="base" value="3.01550069757" unit="amu"/></mass>
                <spin>
                  <fraction label="base" value="1/2" unit="hbar"/></spin>
                <parity>
                  <integer label="base" value="1"/></parity>
                <charge>
                  <integer label="base" value="1" unit="e"/></charge>
                <halflife>
                  <double label="base" value="12.32" unit="yr"/></halflife>
                <energy>
                  <double label="base" value="0." unit="keV"/></energy></nucleus></nuclide></nuclides></isotope></isotopes></chemicalElement>
    <chemicalElement symbol="O" Z="8" name="Oxygen">
      <isotopes>
        <isotope symbol="O16" A="16">
          <nuclides>
            <nuclide id="O16">
              <mass>
                <double label="base" value="15.9949146196" unit="amu"/></mass>
              <nucleus id="o16" index="0">
                <energy>
                  <double label="base" value="0." unit="keV"/></energy></nucleus></nuclide>
            <nuclide id="O16_e1">
              <nucleus id="o16_e1" index="1">
                <energy>
                  <double label="base" value="6049.4" unit="keV"/></energy></nucleus></nuclide>
            <nuclide id="O16_e2">
              <nucleus id="o16_e2" index="2">
                <energy>
                  <double label="base" value="6129.893" unit="keV"/></energy></nucleus></nuclide>
            <nuclide id="O16_e3">
              <nucleus id="o16_e3" index="3">
                <energy>
                  <double label="base" value="6917.1" unit="keV"/></energy></nucleus></nuclide></nuclides></isotope>
        <isotope symbol="O17" A="17">
          <nuclides>
            <nuclide id="O17">
              <mass>
                <double label="base" value="15.9949146196" unit="amu"/></mass>
              <nucleus id="o17" index="0">
                <energy>
                  <double label="base" value="0." unit="keV"/></energy></nucleus></nuclide>
            <nuclide id="O17_e1">
              <nucleus id="o17_e1" index="1">
                <energy>
                  <double label="base" value="6049.4" unit="keV"/></energy></nucleus></nuclide>
            <nuclide id="O17_e2">
              <nucleus id="o17_e2" index="2">
                <energy>
                  <double label="base" value="6129.893" unit="keV"/></energy></nucleus></nuclide>
            <nuclide id="O17_e3">
              <nucleus id="o17_e3" index="3">
                <energy>
                  <double label="base" value="6917.1" unit="keV"/></energy></nucleus></nuclide></nuclides></isotope></isotopes></chemicalElement>
    <chemicalElement symbol="Am" Z="95" name="Americium">
      <isotopes>
        <isotope symbol="Am242" A="242">
          <nuclides>
            <nuclide id="Am242">
              <mass>
                <double label="base" value="242.059549159" unit="amu"/></mass>
              <nucleus id="am242" index="0">
                <spin>
                  <fraction label="base" value="1" unit="hbar"/></spin>
                <parity>
                  <integer label="base" value="-1"/></parity>
                <halflife>
                  <double label="base" value="16.02" unit="h"/></halflife>
                <energy>
                  <double label="base" value="0." unit="keV"/></energy></nucleus></nuclide>
            <nuclide id="Am242_e1">
              <nucleus id="am242_e1" index="1">
                <spin>
                  <fraction label="base" value="0" unit="hbar"/></spin>
                <parity>
                  <integer label="base" value="-1"/></parity>
                <energy>
                  <double label="base" value="44.092" unit="keV"/></energy></nucleus></nuclide>
            <nuclide id="Am242_e2">
              <nucleus id="am242_e2" index="2">
                <spin>
                  <fraction label="base" value="5" unit="hbar"/></spin>
                <parity>
                  <integer label="base" value="-1"/></parity>
                <halflife>
                  <double label="base" value="141." unit="yr"/></halflife>
                <energy>
                  <double label="base" value="48.6" unit="keV"/></energy></nucleus></nuclide>
            <nuclide id="Am242_e3">
              <nucleus id="am242_e3" index="3">
                <spin>
                  <fraction label="base" value="3" unit="hbar"/></spin>
                <parity>
                  <integer label="base" value="-1"/></parity>
                <energy>
                  <double label="base" value="52.7" unit="keV"/></energy></nucleus></nuclide></nuclides></isotope></isotopes></chemicalElement></chemicalElements></PoPs>
